packages/xo-server/src/xo-mixins/metadata-backups.js

```diff
@@ -132,222 +152,383 @@ export default class metadataBackup {
     })
   }
 
-  async _executor({
-    cancelToken,
-    job: job_,
-    logger,
-    runJobId,
-    schedule,
-  }): Executor {
-    if (schedule === undefined) {
-      throw new Error('backup job cannot run without a schedule')
-    }
-
-    const job: MetadataBackupJob = (job_: any)
-    const remoteIds = unboxIdsFromPattern(job.remotes)
-    if (remoteIds.length === 0) {
-      throw new Error('metadata backup job cannot run without remotes')
-    }
-
-    const poolIds = unboxIdsFromPattern(job.pools)
-    const isEmptyPools = poolIds.length === 0
-    if (!job.xoMetadata && isEmptyPools) {
-      throw new Error('no metadata mode found')
-    }
-
+  async _backupXo({ handlers, job, logger, retention, runJobId, schedule }) {
     const app = this._app
-    const { retentionXoMetadata, retentionPoolMetadata } =
-      job?.settings[schedule.id] || {}
 
     const timestamp = Date.now()
-    const formattedTimestamp = safeDateFormat(timestamp)
-    const commonMetadata = {
-      jobId: job.id,
-      jobName: job.name,
-      scheduleId: schedule.id,
-      scheduleName: schedule.name,
-      timestamp,
-    }
-
-    const files = []
-    if (job.xoMetadata && retentionXoMetadata > 0) {
-      const taskId = logger.notice(`Starting XO metadata backup. (${job.id})`, {
-        data: {
-          type: 'xo',
-        },
-        event: 'task.start',
-        parentId: runJobId,
-      })
+    const taskId = logger.notice(`Starting XO metadata backup. (${job.id})`, {
+      data: {
+        type: 'xo',
+      },
+      event: 'task.start',
+      parentId: runJobId,
+    })
 
-      const xoMetadataDir = `${DIR_XO_CONFIG_BACKUPS}/${schedule.id}`
-      const dir = `${xoMetadataDir}/${formattedTimestamp}`
+    try {
+      const scheduleDir = `${DIR_XO_CONFIG_BACKUPS}/${schedule.id}`
+      const dir = `${scheduleDir}/${safeDateFormat(timestamp)}`
 
       const data = JSON.stringify(await app.exportConfig(), null, 2)
       const fileName = `${dir}/data.json`
 
-      const metadata = JSON.stringify(commonMetadata, null, 2)
+      const metadata = JSON.stringify(
+        {
+          jobId: job.id,
+          jobName: job.name,
+          scheduleId: schedule.id,
+          scheduleName: schedule.name,
+          timestamp,
+        },
+        null,
+        2
+      )
       const metaDataFileName = `${dir}/metadata.json`
 
-      files.push({
-        executeBackup: defer(($defer, handler) => {
+      await asyncMap(
+        handlers,
+        defer(async ($defer, handler, remoteId) => {
           $defer.onFailure(() => handler.rmtree(dir))
```

Comment:
Only execute this if the decorated function throws/rejects.

It appears this is no longer possible, what should we do then?

Reply:
We can do it in the catch.